Feature/basic auth

abriosi opened the pull request:
I have added basic auth. 

Keep in mind this is my first attempt at writing go. I'm sure there is a easier way to add this functionality. I tried to make the cleanest possible with zero go experience. That being said I don't except the pull request to be accepted right away.

I will accept every **suggestion**.

For reading and validating basic auth files I used https://github.com/tg123/go-htpasswd

I think https://github.com/abriosi/minikeyvalue/blob/feature/basicAuth/volume#L9-L123 can be improved but my **bash** scripting sucks.

It passes the tests. I think this is the right time to ask for feedback

geohot commented:
At first glance, it has way too many lines and things unrelated to the feature. For example, fix README quotes and spacing in a separate PR first. Why did the number of servers change in the README?

Then, all of these else things aren't good either, multiple code paths. Why don't the tests work with basicauth? Why is that decorator in front of everything.

I'm expecting this feature to add no more than 50 lines, including tests.

geohot commented on tools/test.py:
```diff
@@ -207,6 +258,5 @@ def test_content_hash(self):
-  
-  unittest.main()
 
+  unittest.main()
```
Like why did this change?

geohot commented on tools/rtest.sh:
```diff
@@ -7,16 +7,30 @@ echo "rebuild and rebalance test"
-# rebuild and compare the database
-./mkv -volumes localhost:3001,localhost:3002,localhost:3003,localhost:3004,localhost:3005 -db $ALTDB rebuild
-go run tools/leveldb_compare.go /tmp/indexdb/ $ALTDB
+if [[ -z "${USERPASS}" ]]; then
```
So many better ways to do this. This is unmaintainable.

geohot commented on src/server.go:
```diff
@@ -124,7 +155,7 @@ func (a *App) ServeHTTP(w http.ResponseWriter, r *http.Request) {
       for _, vn := range rand.Perm(len(rec.rvolumes)) {
         remote = fmt.Sprintf("http://%s%s", rec.rvolumes[vn], key2path(key))
-        if remote_head(remote) {
+        if remote_head(remote, authtoken) {
```
Hmm, can you include the basicauth in the remote url?

abriosi replied:
You mean something like this `http://localhost:3001/sv07/60/08/L3dlaGF2ZQ==?authtoken=XXXXXXX`.

And then using a function to transform that into `req.Header.Set("Authorization", authtoken)`?

abriosi replied:
Oh! you mean `http://admin:thisisatest@localhost:3001/sv07/60/08/L3dlaGF2ZQ==`

abriosi commented:
Thanks for the inputs. I have refactored almost everything.

It passes the tests, including a second set I added which tests everything again but with basic auth activated.

Where do you think there is still room for improvement? Perhaps on these `if` bash statements https://github.com/geohot/minikeyvalue/pull/20/files#diff-cab06af6b3f5eed7c24f526b744522b9R5-R12
https://github.com/geohot/minikeyvalue/pull/20/files#diff-78aa2f78ae1dfbe37f302907b253db44R4-R9
https://github.com/geohot/minikeyvalue/pull/20/files#diff-210ab9e731c9c36c2c38db15c28a8d1cR9-R17

In `tests.py` I had to manually handle 302 calls. It seems `requests` strips headers authorization during a 302 because of a random CVE. I'm not completely sure this test is exactly the same https://github.com/geohot/minikeyvalue/pull/20/files#diff-37479d55e637ac7befeefff7f6d88025L122-R137 (I don't understand the original `allow_redirects=True`

geohot commented:
I can already see we are going much better, +144/-33. Reading now, but first, can you think of any more simplifications?

geohot commented on mkv:
```diff
@@ -1,4 +1,5 @@
 #!/bin/bash -e
 go get github.com/syndtr/goleveldb/leveldb
+go get github.com/tg123/go-htpasswd
```
Do we really need this new dependency?

abriosi replied:
Master does a **GET** request with basic authorization to a random volume effectively performing the authorization

geohot commented on src/main.go:
```diff
@@ -99,14 +116,15 @@ func main() {
     http.ListenAndServe(fmt.Sprintf(":%d", *port), &a)
   } else if command == "rebuild" {
-    a.Rebuild()
+    a.Rebuild(userpassValue)
```
Why is this one special, there's tons of other options, and they aren't passed in like this.

geohot commented on src/rebalance.go:
```diff
@@ -12,13 +12,13 @@ type RebalanceRequest struct {
 }
 
-func rebalance(a *App, req RebalanceRequest) bool {
+func rebalance(a *App, req RebalanceRequest, userpass string) bool {
```
Same as above, why isn't this in the App class?

geohot commented on src/server.go:
```diff
@@ -76,9 +77,37 @@ func (a *App) QueryHandler(key []byte, w http.ResponseWriter, r *http.Request) {
   key := []byte(r.URL.Path)
 
+  // check if user is authorized
+  authstring, authorized := a.CheckAuthorization(w, r)
```
Can we outsource the auth check to the volume nodes? I don't like having it in two places

abriosi replied:
Ok, let me see if I understand clearly. You are suggesting that the basic auth is only checked at the nginx, right? I.e., the go server accepts every incoming request, checks with a volume whether that authentication is valid or not, and then proceeds with the logic. That way we don't need any other dependency, and the server is still protected

geohot commented on volume:
```diff
@@ -6,6 +6,16 @@ export PORT=${PORT:-3001}
 mkdir -p $VOLUME
 chmod 777 $VOLUME
 
+if [ "$AUTH" = "" ]
```
Much better on this.

geohot commented on tools/test.py:
```diff
@@ -15,19 +15,32 @@
-    return b"http://localhost:3000/swag-" + binascii.hexlify(os.urandom(10))
+    return b"http://"+str.encode(authstring)+b"localhost:3000/swag-" + binascii.hexlify(os.urandom(10))
+
+  # handle 302 manually https://github.com/psf/requests/issues/2949
```
Hmm, this concerns me a bit about this whole approach. The idea of mkv is that curl and stuff should just work. Is this unique to requests? What does the HTTP spec say?

abriosi replied:
I think they are being too agressive https://github.com/psf/requests/issues/2949#issuecomment-167664262

I will try to check the HTTP spec

abriosi replied:
btw, this is the CVE they are talking about https://cve.mitre.org/cgi-bin/cvename.cgi?name=2014-1829

abriosi replied:
From what I understand requests shouldn't strip the authorization headers according with the HTTP spec https://tools.ietf.org/html/rfc2617#section-3.3

![image](https://user-images.githubusercontent.com/23288462/71495860-ed76db00-2847-11ea-904e-8016ead82cf4.png)

abriosi commented on src/server.go:
```diff
@@ -76,9 +77,38 @@ func (a *App) QueryHandler(key []byte, w http.ResponseWriter, r *http.Request) {
+    }
+  }
+  return authstring, true
+}
```
Nginx performs the basic authorization check before allowing the master to proceed with its operation. It chooses a random volume to perform the verification

geohot commented on README.md:
````diff
@@ -109,3 +113,31 @@ starting thrasher
+```
+
+```
+# pass the .htpasswd to the -auth flag to serve with basic authentification
````
Is this comment right?

abriosi replied:
Corrected

geohot commented on src/main.go:
```diff
@@ -66,8 +68,17 @@ func main() {
   flag.Parse()
 
+  // rebuild of rebalance with basiauth activated on nginx
+  var userpassValue string
```
What is this? I'm confused reading it.

geohot commented on src/rebalance.go:
```diff
@@ -18,7 +18,7 @@ func rebalance(a *App, req RebalanceRequest) bool {
   rvolumes := make([]string, 0)
   for _, rv := range req.volumes {
-    if remote_head(fmt.Sprintf("http://%s%s", rv, kp)) {
+    if remote_head(fmt.Sprintf("http://%s%s%s", a.userpass, rv, kp)) {
```
This change is now on four lines, it should probably be abstracted

abriosi replied:
I tried it by passing userpass into the volumes name, which would mean no change would have to be done to rebalance and rebuild. This doesn't work because userpass ends up being written to levelDB which is unwanted behaviour. (leveldb_compare.go fails)

geohot replied:
Hmm, I see. Okay, this seems fine.

geohot commented on src/server.go:
```diff
@@ -76,9 +77,38 @@ func (a *App) QueryHandler(key []byte, w http.ResponseWriter, r *http.Request) {
 func (a *App) ServeHTTP(w http.ResponseWriter, r *http.Request) {
   key := []byte(r.URL.Path)
 
+  // check if user is authorized
```
I'd put the if a.basicauth in here. It's more readable that way, and GetAuthorization shouldn't have to depend on the app class. Follow principle of least scope.

geohot commented on tools/test.py:
```diff
@@ -119,7 +132,9 @@ def test_head_request(self):
     r = requests.put(key, data=data)
     self.assertEqual(r.status_code, 201)
-    r = requests.head(key, allow_redirects=True)
+    r = requests.head(key, allow_redirects=False)
```
Duplicated code with get_302

geohot commented on src/rebuild.go:
```diff
@@ -127,13 +127,13 @@ func (a *App) Rebuild() {
 
-  parse_volume := func(tvol string) {
-    for _, i := range get_files(fmt.Sprintf("http://%s/", tvol)) {
+  parse_volume := func(tvol string, userpass string) {
```
Can you include the user pass in the volume name?

abriosi replied:
Is this something we really want to do? On L137 we pass `tvol` into `RebuildRequest` which then gets passed into `rebuild` on L123 -> means user pass is going to get written into LevelDB?

geohot replied:
Same as above.

geohot commented on src/server.go:
```diff
@@ -76,9 +77,38 @@ func (a *App) QueryHandler(key []byte, w http.ResponseWriter, r *http.Request) {
+    if authtoken != "" {
+      authstringb, _ := base64.StdEncoding.DecodeString(strings.Split(authtoken, "Basic ")[1])
+      authstring = string(authstringb)
+      authstring = authstring+"@"
```
Put in the sprint

geohot commented on src/server.go:
```diff
@@ -76,9 +77,38 @@ func (a *App) QueryHandler(key []byte, w http.ResponseWriter, r *http.Request) {
+  if a.basicauth == true {
+    authtoken := r.Header.Get("Authorization")
+    if authtoken != "" {
+      authstringb, _ := base64.StdEncoding.DecodeString(strings.Split(authtoken, "Basic ")[1])
```
This can crash

geohot commented on src/server.go:
```diff
@@ -76,9 +77,38 @@ func (a *App) QueryHandler(key []byte, w http.ResponseWriter, r *http.Request) {
+      authstring = authstring+"@"
+      remote := fmt.Sprintf("http://%s%s", authstring, a.volumes[rand.Intn(len(a.volumes))])
+      resp, err := http.Get(remote)
+      if err != nil {
```
Merge these? Why pass through the status code?

abriosi replied:
How do you suggest the merge? If there was an error there is no status code, e.g., error we send out 500; no error we send out nginx status code (such as 401)

geohot replied:
Sure, but I don't like that w and r are being passed in here. Principle of minimal scope.

authtoken := r.Header.Get("Authorization") belongs outside
And I think it should always be a 500 if the volume server is down, so w.WriteHeader(500) belongs on line 115

geohot commented on tools/test.py:
```diff
@@ -15,19 +15,31 @@
+  def handle_302(self, key, headers={}, req_type='GET'):
+    if req_type == 'GET': request_funcion = requests.get
+    if req_type == 'HEAD': request_funcion = requests.head
+    r = request_funcion(key, headers=headers, allow_redirects=False)
```
request_funcion is misspelled. And I don't think the name of this function should be handle_302

geohot commented:
Sorry for the long wait, was busy with work. It's come a long way, few more comments inline.

abriosi commented:
> Sorry for the long wait, was busy with work. It's come a long way, few more comments inline.

No problem George. I have been quite busy (not comma.ai busy though) and I stopped thinking about further enhancements. 

I think this pull request is almost there.

Congratulations on profitable January 🎉🎉

geohot commented:
Just waiting on the GetAuthorization refactor, then I think it's good.

abriosi commented:
Ok, I think everything is set. It passes the tests but I would still like to go through it one more time with a clear head. Tomorrow I will do it

geohot commented:
Sounds good

geohot commented on tools/test.py:
```diff
@@ -15,19 +15,31 @@
+  def custom_request(self, key, headers={}, req_type='GET'):
+    if req_type == 'GET': request_function = requests.get
+    if req_type == 'HEAD': request_function = requests.head
+    r = request_function(key, headers=headers, allow_redirects=False)
```
Multiple redirects still aren't handled!

abriosi replied:
ohh, I see

abriosi commented:
Manual test worked fine. Any other suggestion?

Perhaps get these comments out?
https://github.com/geohot/minikeyvalue/blob/045c7377df7895cac563b8a9a678eebcf36897e0/src/server.go#L88
https://github.com/geohot/minikeyvalue/blob/045c7377df7895cac563b8a9a678eebcf36897e0/src/server.go#L93
https://github.com/geohot/minikeyvalue/blob/045c7377df7895cac563b8a9a678eebcf36897e0/src/server.go#L96

abriosi commented:
Manual testing worker fine. Any other suggestion?

Perhaps remove these comments:
https://github.com/geohot/minikeyvalue/blob/c86281e51a8144f1e0445a6698651df0bc815b73/src/server.go#L88-L96

geohot commented:
All looks good, merging. Congrats!

geohot commented on tools/test.py:
```diff
@@ -15,19 +15,31 @@
+    if req_type == 'GET': request_function = requests.get
+    if req_type == 'HEAD': request_function = requests.head
+    r = request_function(key, headers=headers, allow_redirects=False)
+    while r.status_code == 302:
```
Yes
